jwql/instrument_monitors/common_monitors/readnoise_monitor.py

```diff
@@ -0,0 +1,625 @@
+        current_date = datetime.datetime.now()
+        parameters['TIME-OBS'] = current_date.time().isoformat()
+
+        return parameters
+
+    def make_histogram(self, data):
+        """Creates a histogram of the input data and returns the bin centers 
+        and the counts in each bin.
+
+        Parameters
+        ----------
+        data : numpy.ndarray
+            The input data.
+
+        Returns
+        -------
+        n : numpy.ndarray
+            The counts in each histogram bin.
+
+        bin_centers : numpy.ndarray
+            The histogram bin centers.
+        """
+
+        # Calculate the histogram range as that within 5 sigma from the mean
+        data = data.flatten()
+        clipped = sigma_clip(data, sigma=3.0, maxiters=5)
+        mean, stddev = np.nanmean(clipped), np.nanstd(clipped)
+        lower_thresh, upper_thresh = mean - 4 * stddev, mean + 4 * stddev
+
+        # Some images, e.g. readnoise images, will never have values below zero
+        if (lower_thresh < 0) & (len(data[data < 0]) == 0):
+            lower_thresh = 0.0
+
+        # Make the histogram
+        n, bin_edges = np.histogram(data, bins='auto', range=(lower_thresh, upper_thresh))
+        bin_centers = (bin_edges[:-1] + bin_edges[1:]) / 2
+
+        return n, bin_centers
+
+    def make_readnoise_image(self, data):
+        """Calculates the readnoise for the given input dark current ramp.
+
+        Parameters
+        ----------
+        data : numpy.ndarray
+            The input ramp data. The data shape is assumed to be a 4D array in
+            DMS format (integration, group, y, x).
+
+        Returns
+        -------
+        readnoise : numpy.ndarray
+            The 2D readnoise image.
+        """
+
+        # Create a stack of CDS images (group difference images) using the input ramp data,
+        # combining multiple integrations if necessary.
+        logging.info('\tCreating stack of CDS difference frames')
+        n_ints, n_groups, n_y, n_x = data.shape
+        for integration in range(n_ints):
+            if n_groups % 2 == 0:
+                cds = data[integration, 1::2, :, :] - data[integration, ::2, :, :]
+            else:
+                # Omit the last group if the number of groups is odd
+                cds = data[integration, 1::2, :, :] - data[integration, ::2, :, :][:-1]
+
+            if integration == 0:
+                cds_stack = cds
+            else:
+                cds_stack = np.concatenate((cds_stack, cds), axis=0)
+
+        # Calculate the readnoise by taking the clipped stddev through the CDS stack
+        logging.info('\tCreating readnoise image')
+        clipped = sigma_clip(cds_stack, sigma=3.0, maxiters=3, axis=0)
+        readnoise = np.std(clipped, axis=0)
+        readnoise = readnoise.filled(fill_value=np.nan)  # converts masked array to normal array and fills missing data
+
+        return readnoise
+
+    def most_recent_search(self):
```

Comment:
Hmm maybe, I hesitate though because I'm not sure how generalized the other monitors are for how/if they handle their query history entries, column names/end time units, etc.

Reply:
Ok.  Perhaps something to keep in mind for a future PR.